sdk/core/core/src/az_json_builder.c

```diff
@@ -28,6 +31,80 @@ AZ_NODISCARD static az_result az_json_builder_write_str(
   return AZ_OK;
 }
 
+typedef struct {
+  az_span_action write;
+  az_span value;
+  size_t begin;
+  size_t i;
+} az_json_value_span_state;
+
+AZ_NODISCARD az_result az_json_value_span_state_flush(az_json_value_span_state * const state) {
+  AZ_CONTRACT_ARG_NOT_NULL(state);
+
+  az_span const span = az_span_sub(state->value, state->begin, state->i);
+  if (!az_span_is_empty(span)) {
+    return az_span_action_do(state->write, span);
+  }
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_json_value_span_state_replace(
+    az_json_value_span_state * const state,
+    az_span const replacement) {
+  AZ_RETURN_IF_FAILED(az_json_value_span_state_flush(state));
+  AZ_RETURN_IF_FAILED(az_span_action_do(state->write, replacement));
+  size_t const new_i = state->i + 1;
+  state->i = new_i;
+  state->begin = new_i;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_json_builder_write_span(az_json_builder * const self, az_span const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_json_value_span_state state = {
+    .write = self->write,
+    .value = value,
+    .begin = 0,
+    .i = 0,
+  };
+  AZ_RETURN_IF_FAILED(az_span_action_do(state.write, AZ_STR("\"")));
+  while (true) {
+    az_result_byte const c = az_span_get(state.value, state.i);
+    if (c == AZ_ERROR_EOF) {
+      break;
+    }
+    // check if the character has to be escaped.
+    {
+      az_span const esc = az_json_esc_encode(c);
+      if (!az_span_is_empty(esc)) {
+        AZ_RETURN_IF_FAILED(az_json_value_span_state_replace(&state, esc));
+        continue;
+      }
+    }
+    // check if the character has to be escaped as a UNICODE escape sequence.
+    if (0 <= c && c < 0x20) {
+      uint8_t const array[6] = {
+        '\\',
+        'u',
+        '0',
+        '0',
+        az_number_to_upper_hex((uint8_t)(c / 16)),
+        az_number_to_upper_hex((uint8_t)(c % 16)),
+      };
+      AZ_RETURN_IF_FAILED(
+          az_json_value_span_state_replace(&state, (az_span)AZ_SPAN_FROM_ARRAY(array)));
+      continue;
+    }
+    // no need to escape, process to the next character.
+    state.i += 1;
+  }
+  AZ_RETURN_IF_FAILED(az_json_value_span_state_flush(&state));
+  AZ_RETURN_IF_FAILED(az_span_action_do(state.write, AZ_STR("\"")));
+  return AZ_OK;
+}
+
 // 2^53 - 1
 #define AZ_JSON_UINT_MAX 0x1FFFFFFFFFFFFFull
```

Comment:
unrelated, but I think that using lowercase for long constant will eventually make a problem when we have a constant of `10l`

Reply:
The only problem is uppercase hex letters `0xFFFFULL`.